man/man1/raxpol_file_hdr.1
.\" 
.\" Copyright (c) 2016, Gordon D. Carrie. All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 
.\"     * Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\"     * Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
.\" TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.\" Please address questions and feedback to dev0@trekix.net
.\"
.Pp
.Dd $Mdocdate$
.Dt RAXPOL_FILE_HDR 1
.Os UNIX
.Sh NAME
.Nm raxpol_file_hdr
.Nd Print file header from a RaXPol moment file to standard output.
.Sh SYNOPSIS
.Nm raxpol_file_hdr
.Op Fl V
.Op Fl l
.Op Ar raxpol_file
.Sh DESCRIPTION
With
.Fl V ,
this application just prints version information and exits.
Otherwise, this application prints the file header from a RaXPol moment file to
standard output. If then prints ray headers for the first and last rays in the
format used by
.Nm raxpol_ray_hdrs .
See
.Xr raxpol_ray_hdrs 1 .
If
.Ar raxpol_file
is absent or
.Ql - ,
read standard input.
.Fl l
indicates file uses legacy, pre 2011, format.
.Sh OUTPUT FORMAT
Output has form:
.Bd -literal -offset indent
\fB************************ file header ************************\fP
\fBversion_code\fP \fIinteger\fP
\fBasp_chirp_bandwidth\fP \fIfloat\fP
\fBasp_chirp_center_freq\fP \fIfloat\fP
\fBasp_chirp_intrl_delay\fP \fIinteger\fP
\fBasp_chirp_width\fP \fIfloat\fP
\fBasp_hop_ctr_freq\fP \fIfloat\fP
\fBasp_hop_step_freq\fP \fIfloat\fP
\fBasp_hop_width\fP \fIinteger\fP
\fBasp_num_hops\fP \fIinteger\fP
\fBasp_trigger_delay\fP \fIinteger\fP
\fBasp_tukey_alpha\fP \fIfloat\fP
\fBasp_waveform_type\fP \fIinteger\fP
\fBasp_chirp_delay\fP \fIinteger\fP
\fBclutter_filter_width\fP \fIfloat\fP
\fBclutter_avg_intvl\fP \fIinteger\fP
\fBdata_rate\fP \fIfloat\fP
\fBfft_length\fP \fIinteger\fP
\fBfft_window_type\fP \fIinteger\fP
\fBauto_file_roll_num_rec\fP \fIinteger\fP
\fBauto_file_roll_num_scans\fP \fIinteger\fP
\fBauto_file_roll_flsz\fP \fIinteger\fP
\fBauto_file_roll_start_time\fP \fIinteger\fP
\fBauto_file_roll_type\fP \fIinteger\fP
\fBfilter_bandwidth\fP \fIfloat\fP
\fBgroup_interval\fP \fIinteger\fP
\fBH_dBZ_dBm_offset\fP \fIfloat\fP
\fBH_noise_power\fP \fIfloat\fP
\fBintegration_time\fP \fIfloat\fP
\fBmax_sampled_range\fP \fIfloat\fP
\fBnum_rng_gates\fP \fIcount\fP
\fBnumber_group_pulses\fP \fIinteger\fP
\fBpost_decimation_level\fP \fIinteger\fP
\fBpost_averaging_interval\fP \fIinteger\fP
\fBpri1\fP \fIinteger\fP
\fBpri2\fP \fIinteger\fP
\fBpri_total\fP \fIinteger\fP
\fBprimary_decimation_CIC\fP \fIinteger\fP
\fBpulse_compression\fP \fIflag\fP
\fBfir_output_len_ratio\fP \fIfloat\fP
\fBfir_output_offset_ratio\fP \fIfloat\fP
\fBrange_gate_spacing\fP \fImeters\fP
\fBrange_resolution\fP \fImeters\fP
\fBrecord_moments\fP \fIflag\fP
\fBrecord_raw\fP \fIflag\fP
\fBrecording_enabled\fP \fIflag\fP
\fBservmode\fP \fIcode\fP \fI(name)\fP
\fBserver_state\fP \fIinteger\fP
\fBskip_count\fP \fIinteger\fP
\fBsoftware_decimation_level\fP \fIinteger\fP
\fBsumpower\fP \fIinteger\fP
\fBswitch_delay\fP \fIinteger\fP
\fBswitch_padding\fP \fIinteger\fP
\fBswitch_width\fP \fIinteger\fP
\fBtotal_avg_intvl\fP \fIinteger\fP
\fBtwta_internal_delay\fP \fIinteger\fP
\fBtx_pulse_bracketing\fP \fIinteger\fP
\fBtx_delay\fP \fIinteger\fP
\fBtx_trigger_delay\fP \fIinteger\fP
\fBtx_pulse_center_offset\fP \fIinteger\fP
\fBunambiguous_range\fP \fIkilometers\fP
\fBclutter_filter\fP \fIinteger\fP
\fBcustom_TX_waveform_was_used\fP \fIinteger\fP
\fBV_dBZ_dBm_offset\fP \fIfloat\fP
\fBv_noise_dBm\fP \fIfloat\fP
\fBzero_range_gate_index\fP \fIfloat_meters\fP
\fBscan_type\fP \fIcode\fP \fI(name)\fP
\fBnum_sweeps\fP \fIfloat\fP
\fBtimestamp_sec\fP \fIinteger\fP
\fBtimestamp_usec\fP \fIinteger\fP
\fBpulse_filter_file_name\fP \fIstring\fP
\fBcustom_waveform_file_name\fP \fIstring\fP
\fBnum_rays\fP \fIcount\fP
\fB********************* first ray *********************\fP

.Ed
Then process prints headers for first ray, as described in
.Xr raxpol_ray_hdrs 1 .
Then it prints
.Bd -literal -offset indent
\fB********************** last ray *********************\fP

.Ed
followed by headers for the last ray in the file.
.Sh ENVIRONMENT
Non-zero
.Ev RAXPOL_OLD_FMT
environment variable is same as
.Fl l .
.Sh AUTHOR
.An Gordon Carrie
.Mt dev0@trekix.net


RELEASE_NOTES
Wed Feb 11 16:19:57 CST 2015
raxpol_ray_hdrs also prints file offset to ray

20150316161719
Re-wrote findswps.
Ray header printout includes date as well as time

20150428170551
findswps --
min_span argument can take separate values for PPI and RHI.
Cleaned out dead and deprecated code from findswps.c.
Reformatted output of findswps command for easier searching.
--
raxpol_lib --
Added blocks for 2011 and older raxpol files.
Reformatted output of header print functions for better compatibility with awk.
Added ability to read SPP Sum Power rays.
Various bug fixes. More checking for NULL pointers in functions.
--
raxpol_file_hdr --
prints headers for first and last ray.
Accepts -o option indicating old format files.
--
raxpol_ray_hdr --
Added options.  -a requests abbreviated output.  -h sets heading, overriding
GPS heading in ray headers.  -o input file is "old" (2011) format. Start and
end ray times desired are now specified with -s and -e options, respectively.
--
raxpol_to_sweep --
Accepts separate minimum spans for PPI and RHI.
Is more organinized.
--
tm_calc library has a resolution parameter
raxpol_to_dorade.awk checked in. Makes partial DORADE sweep right now.
raxpol_ray_hdrs utility accepts a time range.

20150507150001
Added RaXPol_Noise function.
Split RaXPol_ZH and RaXPol_ZV into RaXPol_ZH, RaXPol_ZV, RaXPol_DBZH,
and RaXPol_DBZV.
raxpol_file_hdr utility prints offsets to first and last rays.
raxpol_lib --
Created separate raxpol_z and raxpol_dbz functions instead of doing everything
in raxpol_z.
--
raxpol_dat --
utility only calculates and prints requested fields.
raxpol_dat computes moments locally instead of using functions in raxpol_lib,
which are now redundant. RaXPol moments have too many interdependencies for
modular calculations in the current set up.
--

20150925120929
strlcpy.h --
    Disabled global strlcpy before local declaration. This prevents interference
    from Apple's version of strlcpy in their stupid "standard" library.
--
raxpol.h, raxpol_lib.h, raxpol_dat.c --
    Encapsulated the various server modes and recorded moments into RaXPol_Data
    struct.  Consolidated the input and output functions into a smaller set that
    interfaces with the struct. Rewrote raxpol_dat.c with the new interface.
--

20151005153302
findswps.c --
    Added more tests to still_rhi and still_ppi functions.
    Capitalized file scope variables.
--
raxpol.h, raxpol_lib.c, raxpol_dat.c --
    Functions that convert input fields to output are now methods of RaXPol_Data
    struct instead of global funtions.
--

20151005170735
findswps.c --
    Some global variables are now local.
--

20151006200255
raxpol_lib.c --
    Filled in stub functions for single pulse pair and single pulse pair sum
    power output fields.
--

20151013183955
Checking in sweep_img.c, raxpol_swp_idx, raxpol_swp_dat, and raxpol_swp_dat.awk.
Removed raxpol_to_sweep, replace with raxpol_swp_idx.
Checking in geog_proj.c geog_proj.h get_colors.c get_colors.h.
raxpol_lib.c --
    Abbreviated ray header output now includes truck heading.
--
raxpol_dat.c --
    Application no longer prints file header. Printing ray headers is optional.
    Moment names are printed only if there is more than one moment.
--
findswps.c --
    First sweep is index 0, not 1;
--

20151014134133
geog_proj.c --
    GeogProjLonLatToXY and GeogProjXYToLonLat set return arguments to NAN on
    failure.
--
sweep_img.c --
    Added functions to compute gate corners for filled sweeps.
--

20151014190433
findswps, raxpol_swp_idx --
    Now accepts -y option to specify scan type sought, "ppi", "rhi", or "all".
    raxpol_swp_idx uses it.
--

20151102113355
Checking in geog, tm_calc, pisa, color_legend, raxpol_swps_svg, sweep_limits
applications.
Checking in str library utility.
raxpol_ray_hdrs --
    Added ray count option
--
findswps --
    -y option is now -i. This allows the option to pass in
    raxpol_swps_svg, where -y option specifies a style sheet.
--
raxpol_swp_dat, raxpol_swp_dat.awk, raxpol_swp_idx --
    Removed. Not needed.
--
sweep_img --
    Keys end with ":", like pisa.
--
raxpol_lib.c --
    Abbreviated ray header output includes Julian ray time in addition to
    calendar ray time.
--

20151102171827
raxpol_ray_hrs --
    Just takes ray start index and count, or just prints all rays.
    Time search will be a separate application.
--
raxpol_lib.c --
    Shortened abbreviated header output.
--

20151103180433
raxpol_dat --
    Replaced offset argument with index to first ray.
--
sweep_img --
    (bug fix) Corrected order in which gate corners are printed.
    (bug fix) Failure to convert degrees to radians.
--
raxpol_swps_svg --
    Command line options can specify separate bounds for PPI and RHI.
    Adjusted for new abbreviated header output.
--

20151104170343
unix_defs.h --
    c99 is optional.
--
Makefile, raxpol_script_links.svg --
    Path to share directory is substituted into svg script links file.
--

20151104182919
Checking in share directory and contents.
raxpol_swps_svg --
    Color file is now a color directory. Application searches this directory
    for a file with name based on the data type.
    Several variables can now be set either with environment variable or
    command line option.
--

20151105134619
raxpol_dat --
    Always prints moment names, even for only one moment.
--
raxpol_swps_svg --
    Clarified documentation. Strip out moment name from raxpol_dat output.
    sweep_img does not need it.
--

20151105195015
raxpol_dat --
    Now requires starting ray index and ray count. Output consists of moment name
    followed by data for ray count rays, then next moment name followed by data
    for ray count rays, and so on.
--
raxpol_swps_svg --
    Reads the revised raxpol_dat output. Makes sweeps for several moments in
    one block. Desired performance boost failed to show.
--

20151105195715
raxpol_dat, raxpol_swps_svg --
    Restoring from 20151105134619 release. It is simpler and performs as well

src/raxpol_file_hdr.c
/*
   -	raxpol_file_hdr.c --
   -		Print RaXPol moment file header to standard output.
   -		See raxpol_file_hdr (1).
   .
   .	Copyright (c) 2012, Gordon D. Carrie. All rights reserved.
   .	
   .	Redistribution and use in source and binary forms, with or without
   .	modification, are permitted provided that the following conditions
   .	are met:
   .	
   .	    * Redistributions of source code must retain the above copyright
   .	    notice, this list of conditions and the following disclaimer.
   .
   .	    * Redistributions in binary form must reproduce the above copyright
   .	    notice, this list of conditions and the following disclaimer in the
   .	    documentation and/or other materials provided with the distribution.
   .	
   .	THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   .	"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   .	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   .	A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
   .	HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   .	SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
   .	TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   .	PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   .	LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   .	NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   .	SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
   .
   .	Please send feedback to dev0@trekix.net
 */

#include "unix_defs.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include "raxpol.h"

#define RAXPOL_OLD_FMT "RAXPOL_OLD_FMT"

int main(int argc, char *argv[])
{
    char *argv0;			/* Name of the executable, for error
					   messages. */
    char *raxpol_fl_nm;			/* RaXPol file path */
    FILE *raxpol_fl;			/* RaXPol file */
    struct RaXPol_File_Hdr file_hdr;	/* Header from the RaXPol file */
    off_t o0;				/* Offset to start of first ray */
    off_t o1;				/* Offset at end of file */
    long ray_sz;			/* Size in file of one ray */
    struct RaXPol_Ray_Hdr ray_hdr;	/* Ray header from the RaXPol file */

    argv0 = argv[0];
    if ( atoi(getenv(RAXPOL_OLD_FMT) ? getenv(RAXPOL_OLD_FMT) : "0") ) {
	RaXPol_Old_Fmt();
    }
    if ( argc == 1 ) {
	/* raxpol_file_hdr */
	raxpol_fl_nm = "-";
    } else if ( argc == 2 ) {
	if ( strcmp(argv[1], "-V") == 0 ) {
	    printf("%s\n", RAXPOL_MOMENT_VERSION);
	    exit(EXIT_SUCCESS);
	} else if ( strcmp(argv[1], "-l") == 0 ) {
	    /* raxpol_file_hdr -l */
	    raxpol_fl_nm = "-";
	    RaXPol_Old_Fmt();
	} else {
	    /* raxpol_file_hdr raxpol_file */
	    raxpol_fl_nm = argv[1];
	}
    } else if ( argc == 3 && strcmp(argv[1], "-l") == 0 ) {
	/* raxpol_file_hdr -l raxpol_file */
	RaXPol_Old_Fmt();
	raxpol_fl_nm = argv[2];
    } else {
	fprintf(stderr, "Usage: %s raxpol_file\n", argv0);
	exit(EXIT_FAILURE);
    }
    if ( strcmp(raxpol_fl_nm, "-") == 0 ) {
	raxpol_fl = stdin;
    } else if ( !(raxpol_fl = fopen(raxpol_fl_nm, "r")) ) {
	fprintf(stderr, "%s: could not open %s for reading.\n",
		argv0, raxpol_fl_nm);
	exit(EXIT_FAILURE);
    }
    RaXPol_Init_File_Hdr(&file_hdr);
    if ( !RaXPol_Read_File_Hdr(&file_hdr, raxpol_fl) ) {
	fprintf(stderr, "%s: failed to read file header.\n", argv0);
	exit(EXIT_FAILURE);
    }
    printf("************************ file header ************************\n");
    RaXPol_FPrintf_File_Hdr(&file_hdr, stdout);
    if ( (o0 = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not determine file position after reading "
		"file header.\n", argv0);
	exit(EXIT_FAILURE);
    }
    if ( !RaXPol_Read_Ray_Hdr(&ray_hdr, raxpol_fl) ) {
	fprintf(stderr, "%s: failed to read first ray header.\n", argv0);
	exit(EXIT_FAILURE);
    }
    if ( (o1 = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not determine file position after reading "
		"first ray header.\n", argv0);
	exit(EXIT_FAILURE);
    }
    ray_sz = o1 - o0 + ray_hdr.data_size;
    if ( fseek(raxpol_fl, 0, SEEK_END) == -1
	    || (o1 = ftello(raxpol_fl)) == -1 ) {
	fprintf(stderr, "%s: could not set position for last ray of file.\n",
		argv0);
	exit(EXIT_FAILURE);
    }
    o1 = o0 + ((o1 - o0) / ray_sz) * ray_sz;
    printf("num_rays %zd\n", (size_t)((o1 - o0) / ray_sz));
    printf("********************* first ray *********************\n");
    printf("Offset %lu\n", (unsigned long)o0);
    RaXPol_FPrint_Ray_Hdr(&ray_hdr, stdout);
    printf("********************** last ray *********************\n");
    printf("Offset %lu\n", (unsigned long)o1);
    if ( fseek(raxpol_fl, o1 - ray_sz, SEEK_SET) == -1 ) {
	fprintf(stderr, "%s: could not set position for last ray of file.\n",
		argv0);
	exit(EXIT_FAILURE);
    }
    if ( !RaXPol_Read_Ray_Hdr(&ray_hdr, raxpol_fl) ) {
	fprintf(stderr, "%s: failed to read last ray header.\n", argv0);
	exit(EXIT_FAILURE);
    }
    RaXPol_FPrint_Ray_Hdr(&ray_hdr, stdout);
    return EXIT_SUCCESS;
}

